webauthn-security-key/libs/auth.js

```diff
@@ -0,0 +1,694 @@
+/*
+ * @license
+ * Copyright 2021 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+const express = require('express');
+const router = express.Router();
+const crypto = require('crypto');
+const fido2 = require('@simplewebauthn/server');
+const base64url = require('base64url');
+const fs = require('fs');
+const low = require('lowdb');
+const { check, validationResult } = require('express-validator');
+
+if (!fs.existsSync('./.data')) {
+  fs.mkdirSync('./.data');
+}
+
+const FileSync = require('lowdb/adapters/FileSync');
+const adapter = new FileSync('.data/db.json');
+const db = low(adapter);
+
+router.use(express.json());
+
+db.defaults({
+  users: [],
+}).write();
+
+const RP_NAME = 'webauthn-security-key-codelab';
+const FIDO_TIMEOUT = 30 * 1000 * 60;
+const VALID_TRANSPORTS = Object.freeze([
+  'usb',
+  'ble',
+  'nfc',
+  'internal',
+  'cable',
+]);
+
+// Authentication types
+const authTypes = Object.freeze({
+  SINGLE_FACTOR: 'sfa',
+  TWO_FACTOR: '2fa',
+});
+
+// Authentication statuses
+const authStatuses = Object.freeze({
+  NEED_SECOND_FACTOR: 'needSecondFactor',
+  COMPLETE: 'complete',
+});
+
+// Generic message, because an attacker should not be able to determine that a password was correct by looking at the error messages
+const GENERIC_AUTH_ERROR_MESSAGE =
```

Comment:
WebAuthn question:
What I'm trying to do by using  `GENERIC_AUTH_ERROR_MESSAGE` is  that this error message does not give the attacker a hint on whether or not the password was correct (see where `GENERIC_AUTH_ERROR_MESSAGE ` is 
used below, in `/authenticate-two-factor` and `/credential-options`). 
WDYT?